File: server/controllers/mail.go
package controllers

import (
	"server/models"
	"server/utils"

	"github.com/gofiber/fiber/v2"
)

func SendMail(c *fiber.Ctx) error {
	var mail models.Mail

	if err := c.BodyParser(&mail); err != nil {
		return c.JSON(Response{
			"bad formatting",
			false,
			nil,
		})
	}

	if !utils.CheckEmail(mail.Email) {
		return c.JSON(Response{
			"please enter a valid email",
			false,
			nil,
		})
	}

	if len(mail.Message) < 15 {
		return c.JSON(Response{
			"message must be at least 15 characters long",
			false,
			nil,
		})
	}

	if len(mail.Name) < 2 {
		return c.JSON(Response{
			"please include a name",
			false,
			nil,
		})
	}

	if err := mail.SaveMail(); err != nil {
		return c.JSON(Response{
			"unable to send mail",
			false,
			nil,
		})
	}

	return c.JSON(Response{
		"mail sent",
		true,
		nil,
	})
}

File: server/src/controllers/mail.ts
import { Request, Response } from 'express';
import sgMail from '@sendgrid/mail';

import { env } from '../config';

import { ResponseBody, EmailRequestBody } from './controllers.types';

sgMail.setApiKey(env.SENDGRID_API_KEY);

export const sendMail = async (
	req: Request<undefined, ResponseBody, EmailRequestBody>,
	res: Response
) => {
	const msg = {
		to: env.SENDGRID_SENDER_EMAIL,
		from: env.SENDGRID_SENDER_EMAIL,
		subject: 'Mail from contact form - isakgranqvist.com',
		html: `
            <div>
                <h3>${req.body.email}</h3>
                <h4>${req.body.name}</h4>
                <p>${req.body.message}</p>
            </div>
        `,
	};

	try {
		await sgMail.send(msg);

		return res.json({
			message: 'Message sent',
			success: true,
			data: null,
		});
	} catch {
		return res.json({
			message: 'Server error',
			success: false,
			data: null,
		});
	}
};
